[TensorPipe] Close and join TP context at shutdown

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* **#38934 [TensorPipe] Close and join TP context at shutdown**
* #38933 [TensorPipe] Implement join correctly
* #38930 [TensorPipe] Always complete futures from thread pool
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging



The TensorPipe context contains all the threads and global state. It needs to be closed and joined upon shutdown (joining implicitly closes it). Destructing the context implicitly joins it, which is what was happening so far: we were waiting for the RPC agent to be destroyed for the TP context to be closed. However, I was seeing some TSAN errors that seemed to be happening during the process termination, where the SHM reactor thread was trying to log something on GoogleLog while a static member of GoogleLog was being destructed. I suspect this means that this means that the TP agent was being "leaked" (probably because the `RpcAgent::currentRpcAgent_` static field was still storing it) and thus was destroyed too late. The obvious solution seems to be to destroy it earlier, when GoogleLog is still active.

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703016/)!

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016)

lw committed:
[TensorPipe] Close and join TP context at shutdown

The TensorPipe context contains all the threads and global state. It needs to be closed and joined upon shutdown (joining implicitly closes it). Destructing the context implicitly joins it, which is what was happening so far: we were waiting for the RPC agent to be destroyed for the TP context to be closed. However, I was seeing some TSAN errors that seemed to be happening during the process termination, where the SHM reactor thread was trying to log something on GoogleLog while a static member of GoogleLog was being destructed. I suspect this means that this means that the TP agent was being "leaked" (probably because the `RpcAgent::currentRpcAgent_` static field was still storing it) and thus was destroyed too late. The obvious solution seems to be to destroy it earlier, when GoogleLog is still active.

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703016/)!

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Close and join TP context at shutdown"




The TensorPipe context contains all the threads and global state. It needs to be closed and joined upon shutdown (joining implicitly closes it). Destructing the context implicitly joins it, which is what was happening so far: we were waiting for the RPC agent to be destroyed for the TP context to be closed. However, I was seeing some TSAN errors that seemed to be happening during the process termination, where the SHM reactor thread was trying to log something on GoogleLog while a static member of GoogleLog was being destructed. I suspect this means that this means that the TP agent was being "leaked" (probably because the `RpcAgent::currentRpcAgent_` static field was still storing it) and thus was destroyed too late. The obvious solution seems to be to destroy it earlier, when GoogleLog is still active.

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703016/)!

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -522,17 +522,17 @@ void TensorPipeAgent::join() {
 void TensorPipeAgent::shutdownImpl() {
+  // This will close all the pipes and listeners, invoke all callbacks with
+  // errors, turn down the I/O event loops and wait for everything to terminate.
+  context_->join();
```
Why this is before `threadPool_.waitWorkComplete();` and `join`? What if there are pending sends in the `threadPool_`? Even in the `graceful=False` case (i.e., no `join`), we might want to deliver all messages in the best effort manner.

lw replied:
My reasoning here was that if we did this the other way around, by first waiting on the thread pool, then in some cases we may never shutdown: in particular, if there is a continuous stream of incoming requests, the thread pool would just always be busy and the wait would never return. By first closing the TP context, however, we close all pipes and thus we close the "tap" through which new requests may come in. I do agree however that it makes it much less "graceful". If you think that we should still be somewhat graceful also in this case I'll swap the order.

mrshenli replied:
I see. I don't see a reason for using `graceful=False` in OSS with our current implementation. So, it should be fine to land this as-is.

lw committed:
Update on "[TensorPipe] Close and join TP context at shutdown"




The TensorPipe context contains all the threads and global state. It needs to be closed and joined upon shutdown (joining implicitly closes it). Destructing the context implicitly joins it, which is what was happening so far: we were waiting for the RPC agent to be destroyed for the TP context to be closed. However, I was seeing some TSAN errors that seemed to be happening during the process termination, where the SHM reactor thread was trying to log something on GoogleLog while a static member of GoogleLog was being destructed. I suspect this means that this means that the TP agent was being "leaked" (probably because the `RpcAgent::currentRpcAgent_` static field was still storing it) and thus was destroyed too late. The obvious solution seems to be to destroy it earlier, when GoogleLog is still active.

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703016/)!

Differential Revision: [D21703016](https://our.internmc.facebook.com/intern/diff/D21703016)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -529,10 +529,10 @@ void TensorPipeAgent::shutdownImpl() {
+
+  // This will close all the pipes and listeners, invoke all callbacks with
+  // errors, turn down the I/O event loops and wait for everything to terminate.
+  context_->join();
```
Now I understand your reasoning, and I think either is OK. It's up to you whether it should be in the current state or revert to the previous version.
